Add IPsec feature guide (#236)
ipdk-io/networking-recipe: docs/guides/features/ipsec-guide.md

Reviewer: Oops. One more: could you please rename the file to `ipsec-offload.md`?

Author: Fixed

(IKE protocol) is offloaded, thus avoiding specialized drivers or control
planes burdening compute instances. Further details are available on the
[IPDK Inline Acceleration - IPsec](https://ipdk.io/documentation/Recipes/InlineIPsec/)
page.

This document is intended to be used in conjunction with the
[IPsec Recipe](https://github.com/ipdk-io/ipsec-recipe), which provides
strongSwan plugin code to configure and program control messages into the
target.

The IPsec offload feature is only supported on the Intel&reg; IPU E2100 target.

## Feature overview

The `infrap4d` process provides the gRPC server-side support for P4RT and
gNMI messages. strongSwan acts as the gRPC client in this context.
Alternatively, you can use other clients that implement the IKE stack. The
client communicates using P4RT to program Security Policy Database (SPD) and
uses gNMI to configure Security Association Database (SAD) entries, which
includes encryption keys, re-keying etc.

The [openconfig-ipsec-offload](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang)
YANG model is used to configure IPsec offload.

> 

Reviewer: The switch from header level two to header level three doesn't make sense. The following paragraphs aren't part of the Feature Overview, they represent a new topic, which is a How-To.

My suggestion would be to insert a level two heading at this point.

## Enabling IPsec

Follow the sequence of steps . . .

Follow the instructions in [Deploying P4 programs](/guides/es2k/deploying-p4-programs.md)
to load the hardware FXP pipeline with the crypto package.

### Configure and run infrap4d

The OpenConfig portion of IPsec Offload functionality is enabled by using fixed
functions configuration. Follow the instructions in [Running infrap4d](/guides/es2k/running-infrap4d.md)
to prepare system with generated TDI.json and context.json file references.

The /usr/share/stratum/es2k/es2k_skip_p4.conf file must include the fixed
function configuration reference.

```json
"fixed_functions" : [
  {
   "name": "crypto",
   "tdi": "/tmp/fixed/tdi.json",
   "ctx": "/tmp/fixed/crypto-mgr-ctx.json"
  }
],
```

With this configuration updated, you can start infrap4d.

### Compile P4 program

Compile the ipsec-offload program according to the instructions in
[Compiling P4 programs](/guides/es2k/compiling-p4-programs.md)
to generate crypto P4 artifacts for programming the pipelines.

### Configure strongSwan

Follow the instructions in the [strongSwan documentation](https://docs.strongswan.org/docs/5.9/index.html)
to configure host for the use-case selected. This includes details such as

> certificate location, IPs, lifetime thresholds etc.

Reviewer: ```suggestion
certificate location, IPs, lifetime thresholds, etc.
```
_etc._ is part of the list.

Author: hmmm good point. I seem to have forgotten about this over the years.

### Start IPsec

> With the strongSwan application configured, starting IPsec

Reviewer: I would suggest:

"With the strongSwan application configured, starting the [IPsec Recipe](https://github.com/ipdk-io/ipsec-recipe) will initiate the pipeline . . ."

This makes the link an integral part of the narrative, rather than an intrusion.

(see [ipsec-recipe](https://github.com/ipdk-io/ipsec-recipe) for details) will 
initiate the pipeline, program the SPD rules as per the P4 program, and
configure/re-configure SAD entries based on negotiated encryption parameters
between local and peer system.

## OpenConfig and gNMI messages

This section provides detailed information on OpenConfig model and gNMI
messages with the expected format. The strongSwan plugin has the following
details encoded. This is only provided for clarity and information purposes.

> ### The Config SAD message

Reviewer: ```suggestion
### Config SAD message
```

> The P4 Control Plane provides SET and DELETE API calls service to the 

Reviewer: ```suggestion
The SET and DELETE API calls . . .
```
You're describing the individual messages. They (not P4 Control Plane) are your topic.

"The GET and DELETE API calls service to the IPsec-offload Config SAD message at . . ." doesn't make sense.

Perhaps something along the lines of:

"The SET and DELETE API calls are used to configure the Security Association Database. See the IPsec-offload [Config SAD message](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang#L39-L185) ..."

IPsec-offload [Config SAD message](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang#L39-L185)
at `/ipsec-offload/sad/sad-entry[offload-id=x][direction=y]/config`.

Note that individual leaf nodes of the YANG tree cannot be set and the gRPC
server expects the full config message as one single message. This design
was chosen in order to avoid caching of the encryption keys in the
software stack.

The expected format of the Config SAD message is:

```text
offload_id: 1,
direction: true,
req_id: 2,
spi: 599,
ext_seq_num: false,
anti_replay_window_size: 64,
protocol_parameters: IPSEC_PROTOCOL_PARAMS_ESP,
mode: IPSEC_MODE_TRANSPORT,
esp_payload {
  encryption {
    encryption_algorithm: 12,
    key: "33:65:6b:12:d7:9f:56:63:0b:59:a3:9f:8b:03:d2:c1:b8:51:98:e8",
    key_len: 20,
  }
},
sa_hard_lifetime {
  bytes: 200000
},
sa_soft_lifetime {
  bytes: 100000
}
```

> ### The Fetch SPI message

Reviewer: ```suggestion
### Fetch SPI message
```

> The P4 Control Plane supports a GET gNMI call to retrieve the

Reviewer: ```suggestion
The GET gNMI call is used to retrieve the
```

[SPI value](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang#L292)
at `/ipsec-offload/ipsec-spi/rx-spi`.

> ### The Key Expiry Notification message

Reviewer: ```suggestion
### Key Expiry Notification message
```

> The P4 Control Plane supports a [gRPC Notification message](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang#L308) at `/ipsec-offload`. This is used as a signal to trigger the

Reviewer: ```suggestion
The [gRPC Notification message](https://github.com/ipdk-io/openconfig-public/blob/master/release/models/ipsec/openconfig-ipsec-offload.yang#L308) at `/ipsec-offload` is used as a signal to trigger the
```
